### PR TITLE
Add `static_shape`

### PR DESCRIPTION
**Pull request by adityagoel4512**
Possible solution to https://github.com/Quantco/ndonnx/issues/35

### CODE REVIEW

File: ndonnx/_array.py [hunk 1]
```diff
@@ -371,8 +368,8 @@ def __bool__(self) -> bool:
             raise ValueError(f"Cannot convert Array of shape {self.shape} to a bool")
 
     def __len__(self) -> int:
-        if isinstance(eager_value := self.shape, tuple):
-            return eager_value[0]
+        if isinstance(self.shape[0], int):
+            return self.shape[0]
         else:
             raise ValueError(f"Cannot convert Array of shape {self.shape} to a length")
```
Comment (neNasko1) [inline ndonnx/_array.py:374]: This error is not very coherent, can we change it?

File: ndonnx/_core/_shapeimpl.py [hunk 2]
```diff
@@ -24,6 +25,15 @@ def shape(self, x):
             current = next(iter(current._fields.values()))
         return from_corearray(opx.shape(current))
 
+    def static_shape(self, x) -> tuple[int | None, ...]:
+        current = x
+        while isinstance(current, ndx.Array):
```
Comment (cbourjau) [inline ndonnx/_core/_shapeimpl.py:30]:
```suggestion
        # FIXME: Implicit assumption that all fields represent the shape of `x` correctly
        while isinstance(current, ndx.Array):
```

File: ndonnx/additional/__init__.py [hunk 3]
```diff
@@ -1,12 +1,13 @@
 # Copyright (c) QuantCo 2023-2024
 # SPDX-License-Identifier: BSD-3-Clause
 
-from ._additional import fill_null, make_nullable, isin, shape, static_map
+from ._additional import fill_null, make_nullable, isin, shape, static_map, static_shape
```
Comment (cbourjau) [inline ndonnx/additional/__init__.py:4]: Do we need to export `static_shape` publicly? When would one use it rather than `Array.shape`?

**Review by cbourjau — COMMENTED** [review]
Thanks for taking this on! Is this causing significant churn in our dependent projects?

File: ndonnx/_array.py (hunk 1 above)
Comment (adityagoel4512) [reply ndonnx/_array.py:374]: Good point, and neither are `__bool__` and `__int__` error messages.

File: ndonnx/_core/_shapeimpl.py (hunk 2 above)
Comment (adityagoel4512) [reply ndonnx/_core/_shapeimpl.py:30]: We are in `_shapeimpl.py` which is used in dtypes that satisfy that condition - see https://github.com/Quantco/ndonnx/blob/main/ndonnx/_core/_shapeimpl.py#L17-L19.

File: ndonnx/additional/__init__.py (hunk 3 above)
Comment (adityagoel4512) [reply ndonnx/additional/__init__.py:4]: I considered leaving it private but I think the convention of having all overridable functionality being standalone public functions is valuable when contributing custom dtypes. On second thought we can always make it public later so I'll make it private for this release and then revisit this later.

**Comment by adityagoel4512** [comment]
> Is this causing significant churn in our dependent projects?

Nothing very significant. The only thing that changes is that doing ``ndx.asarray(x.shape)`` becomes ``nda.shape(x)``. This is already the case usually and technically we always had the liberty to return `None` dimensions.

**Review by cbourjau — APPROVED** [review]

**Pull request merged**